frontend/src/frontend_benchmark/source_generator.py

```diff
@@ -0,0 +1,185 @@
+            f'{variable} = strtoul(optarg, NULL, 0);\n'
+            '}\n'
+            '}'
+        )
+        return template
+
+    def append_path_to_output_dir(self, path) -> str:
+        return os.path.join(self.output_dir, path)
+
+    def write_headers(self) -> None:
+        header_path = self.append_path_to_output_dir(self.header_file)
+        with open(header_path, 'w') as f:
+            f.write(self.callgraph.format_vars_declaration())
+            f.write(self.callgraph.format_headers())
+
+    def write_functions(self, splits: int = None) -> Collection[str]:
+        grouper = FileFunctionMapper(self.callgraph)
+        file_to_function_name_mapping = grouper.group_functions(splits)
+        for function_file, functions in file_to_function_name_mapping.items():
+            self.write_header_import(function_file)
+            for function_name in functions:
+                self.write_function(function_name, function_file)
+        return file_to_function_name_mapping.keys()
+
+    def write_header_import(self, path) -> None:
+        path = self.append_path_to_output_dir(path)
+        header = self.get_header_import_string()
+        with open(path, 'a') as f:
+            f.write(f'{header}\n\n')
+
+    def write_function(self, function_name: int, path: str) -> None:
+        path = self.append_path_to_output_dir(path)
+        string = self.callgraph.format_function(function_name)
+        with open(path, 'a') as f:
+            f.write(string + '\n')
+
+    def write_makefile(self, function_files: Collection[str]) -> None:
+        makefile_path = self.append_path_to_output_dir('Makefile')
+        with open(makefile_path, 'w') as f:
+            dependencies = self.get_object_files_to_c_files_mapping(function_files)
+            obj_files = ' '.join(dependencies.keys())
+            string = f'{self.benchmark_name}: {obj_files}\n'
+            string = f'{string}\tgcc -o {self.benchmark_name} {obj_files} -O0\n\n'
+            for obj_file, c_file in dependencies.items():
+                string += f'{obj_file}: {c_file}\n'
+                string += f'\tgcc -c -o {obj_file} {c_file} -O0\n\n'
+            string += f'clean:\n\trm *.o {self.benchmark_name}\n'
+            f.write(string)
+
+    def get_object_files_to_c_files_mapping(self, c_files: Collection[str]) -> Dict[str, str]:
+        result = {}
+        for func_file in c_files:
+            obj_file = re.sub(r'\.c', r'.o', func_file)
+            result[obj_file] = func_file
+        result['main.o'] = 'main.c'
+        return result
+
+
+class FileFunctionMapper:
+    def __init__(self, callgraph):
+        self.function_files = set()
+        self.callgraph = callgraph
+
+    def group_functions(self, splits: int = None):
+        result = self._group_functions_without_splits()
+        if splits:
+            result = self._group_functions_with_splits(splits, result)
+        return result
+
+    def _group_functions_without_splits(self) -> Dict[str, Set[int]]:
```

Comment:
I think the purpose of the grouping here is to group all functions that branch/call to each other in a single file? If so, a comment would be helpful. 

I'm also curious how this works for something like the DFS benchmark. Regardless of how deep that tree is, wouldn't you always generate one grouping, since everything is connected to each other?

Reply:
> I think the purpose of the grouping here is to group all functions that branch/call to each other in a single file? If so, a comment would be helpful.

Yes that is right. Will add a comment.

> I'm also curious how this works for something like the DFS benchmark. Regardless of how deep that tree is, wouldn't you always generate one grouping, since everything is connected to each other?

Yes in that case only one grouping is generated. The splits parameter was intended to ease that case. In my tests, I set splits to the same number of CPUs online.